# Practicing Python Collections

In [16]:
import json

In [17]:
with open('../../data/players_data.json') as fp:
    players = json.load(fp)

In [18]:
columns = [
    'Player', 'Nation', 'Pos', 'Squad', 'Comp', 'Age', 'Born', 'MP', 'Starts', 'Min', 'Goals', 'Shots', 'SoT', 'Assists'
]

Get the details of first player

In [19]:
print(players[0])

['Brenden Aaronson', 'USA', 'MFFW', 'Leeds United', 'Premier League', 22, 2000, 20, 19, 1596, 1, 1.53, 0.28, 0.11]


The list is ['Brenden Aaronson', 'USA', 'MFFW', 'Leeds United', 'Premier League', 22, 2000, 20, 19, 1596, 1, 1.53, 0.28, 0.11] The details for the player "Brenden Aaronson" are as follows:
| Number | Field | Information | Comment |
|--------|-------|-------------|---------|
|1.  | Player | Brenden Aaronson |        # Name of the player |
|2.  | Nation | USA              |       # Nationality of the player |
|3.  | Pos | MFFW                |       # Position of the player |
|4.  | Squad | Leeds United      |       # Club/Squad of the player |
|5.  | Comp | Premier League     |       # Competition of the player |
|6.  | Age | 22                  |       # Age of the player |
|7.  | Born | 2000               |       # Birth year of the player |
|8.  | MP | 20                   |       # Matches Played |
|9.  | Starts | 19               |       # Matches Started |
|10. | Min | 1596                |      # Minutes Played |
|11. | Goals | 1                 |      # Goals Scored |
|12. | Shots | 1.53              |      # Shots per 90 minutes |
|13. | SoT | 0.28                |      # Shots on Target per 90 minutes |
|14. | Assists | 0.11            |      # Assists per 90 minutes |

Similarly, the other entries in the list follow the same format, providing information about different players and their performance statistics.

## Analytical Questions



### 1. Who's the player with most assist in the Premier League?
Find the player with the highest number of assists that played AT LEAST1000 minutes. The assists value is computed "per minute". Provide the full name of the player, followed by their assists number. Only consider players who have played at least 1000 minutes.

In [20]:
top_player = ""
assist_stat = 0
for player in players:
    if player[4] == 'Premier League' and player[9] > 1000 and player[13] > assist_stat:
        top_player = player[0]
        assist_stat = player[13]
top_player, assist_stat

('Kevin De Bruyne', 0.62)

### 2. Find the player with the highest goals-to-shots ratio in the Premier League. Only consider players who have taken at least 10 shots.

Only consider players:

- who have taken at least 10 shots per 90 minutes( as shots per 90 minutes is a normalized value, it is not the total number of shots taken by the player in the season so, we need to consider only those players who have taken at least 10 shots per 90 minutes not the total number of shots taken by the player in the season)
- played in the Premier League

In [21]:
top_player = ""
top_ratio = 0
for player in players:
    name = player[0]
    comp = player[4]
    shots = player[11]
    goals = player[10]
    if comp == 'Premier League' and shots >= 10:
        print(name, comp, shots, goals)
        try:
            ratio = goals / shots
        except ZeroDivisionError:
            ratio = 0
            
        if ratio > top_ratio:
            top_ratio = ratio
            top_player = name
top_player, top_ratio


Nathaniel Chalobah Premier League 10.0 0
Halil Derviþoðlu Premier League 10.0 0
João Gomes Premier League 10.0 1
Darko Gyabi Premier League 10.0 0
Djed Spence Premier League 10.0 0


('João Gomes', 0.1)

### 3. Determine the average age of players in each squad.

Create a dictionary named average_age_per_squad where the keys are the squad names, and the values are the average ages rounded to the nearest whole number.

In [22]:
squad_age_count = {}

for player in players:
    squad_name = player[3]
    age = player[5]
    if squad_name not in squad_age_count:
        squad_age_count[squad_name] = {'sum_age': age, 'count': 1}
    else:
        squad_age_count[squad_name]['sum_age'] += age
        squad_age_count[squad_name]['count'] += 1

average_age_per_squad = {key : round(values['sum_age'] / values['count']) for key, values in squad_age_count.items()}
average_age_per_squad

{'Leeds United': 25,
 'Reims': 23,
 'Angers': 26,
 'Lens': 26,
 'Lorient': 24,
 'Hellas Verona': 25,
 'Auxerre': 27,
 'Rennes': 24,
 'Betis': 28,
 'Toulouse': 24,
 'Roma': 26,
 'Inter': 28,
 'Nantes': 26,
 'Sevilla': 27,
 'Southampton': 25,
 'Köln': 25,
 'Fulham': 27,
 'Dortmund': 25,
 'Leverkusen': 25,
 'Milan': 26,
 'Bologna': 26,
 'Cremonese': 26,
 'Troyes': 25,
 'Werder Bremen': 26,
 'Valladolid': 27,
 'Spezia': 25,
 'West Ham': 27,
 'Monaco': 24,
 'Crystal Palace': 26,
 'Stuttgart': 24,
 'Strasbourg': 26,
 'Celta Vigo': 26,
 'Torino': 24,
 'Wolves': 26,
 'Brentford': 25,
 'Mainz 05': 25,
 'Manchester City': 26,
 'Almería': 26,
 'Empoli': 24,
 'Hoffenheim': 25,
 'Real Madrid': 28,
 'Cádiz': 28,
 'Eint Frankfurt': 26,
 'Barcelona': 26,
 'Lazio': 27,
 'Villarreal': 27,
 'Leicester City': 26,
 'Liverpool': 26,
 'Getafe': 27,
 'Elche': 28,
 'Real Sociedad': 24,
 'Clermont Foot': 27,
 'Everton': 26,
 'Valencia': 24,
 'Newcastle Utd': 27,
 'Ajaccio': 27,
 'Schalke 04': 27,
 'Sassuolo': 2

### 4. Which player has the highest playing time in the Premier League?

Find the player who has played the most minutes in Premier League.

In [23]:
player_most_played = None
max_play_time = 0
for player in players:
    name = player[0]
    time = player[9]
    corp = player[4]
    if corp != 'Premier League':
        continue
    else:
        if time > max_play_time:
            max_play_time = time
            player_most_played = name
player_most_played

'Tim Ream'

### 5. What is the highest number of minutes played by a player in La Liga?

Find the maximum minutes played in La Liga. Here La Liga is the Premier League/Comeptition name.

In [24]:
max_time = 0
for player in players:
    corp = player[4]
    time_played = player[9]
    if corp != 'La Liga':
        continue
    else:
        if time_played > max_time:
            max_time = time_played
max_time        

1890

### 6. Which team, on average, scores the highest number of goals per game in the Bundesliga?

Consider only teams that:

- where players have played at least 10 matches
- are in the Bundesliga competition.

Average Goals per Game is defined as: the total goals scored divided by the number of games played by the team.

Find the highest average goals per game and the name of the team which scored the highest goals per game

In [25]:
qualified_teams = {}
for player in players:
    squad_name = player[3]  # Index 3 corresponds to the 'Squad' column
    matches_played = player[7]  # Index 7 corresponds to the 'MP' column
    goals_scored = player[10]  # Index 10 corresponds to the 'Goals' column

    if player[4] == 'Bundesliga' and matches_played >= 10:
        if squad_name in qualified_teams:
            qualified_teams[squad_name]['matches_played'] += matches_played
            qualified_teams[squad_name]['goals_scored'] += goals_scored
        else:
            qualified_teams[squad_name] = {'matches_played': matches_played, 'goals_scored': goals_scored}

# Calculate the average goals per game for each qualified team
average_goals_per_game = {
    squad: data['goals_scored'] / data['matches_played'] for squad, data in qualified_teams.items()
}

# Find the team with the highest average goals per game
highest_average_goals = max(average_goals_per_game.values())
highest_average_goals_team = [squad for squad, avg_goals in average_goals_per_game.items() if avg_goals == highest_average_goals][0]

print("Highest Average Goals Per Game:", highest_average_goals)
print("Team with Highest Average Goals Per Game:", highest_average_goals_team)

Highest Average Goals Per Game: 0.20863309352517986
Team with Highest Average Goals Per Game: Bayern Munich


## Data Transformation

### 7. Transform the list of lists into a list of dictionaries

Create a new variable named players_dict that contains ALL the players but each player represented as a dictionary, containing only the keys:

- player
- nation
- position
- squad
- competition
- age

The resulting variable should have this structure:
<code>
[{'player': 'Brenden Aaronson',
  'nation': 'USA',
  'position': 'MFFW',
  'squad': 'Leeds United',
  'competition': 'Premier League',
  'age': 22},
 {'player': 'Yunis Abdelhamid',
  'nation': 'MAR',
  'position': 'DF',
  'squad': 'Reims',
  'competition': 'Ligue 1',
  'age': 35}]
</code>

In [26]:
players_dict = []
for player in players:
    player_data = {
        'player': player[0],       # Index 0 corresponds to the 'Player' column
        'nation': player[1],       # Index 1 corresponds to the 'Nation' column
        'position': player[2],     # Index 2 corresponds to the 'Pos' column
        'squad': player[3],        # Index 3 corresponds to the 'Squad' column
        'competition': player[4],  # Index 4 corresponds to the 'Comp' column
        'age': player[5],          # Index 5 corresponds to the 'Age' column
    }
    players_dict.append(player_data)

players_dict

[{'player': 'Brenden Aaronson',
  'nation': 'USA',
  'position': 'MFFW',
  'squad': 'Leeds United',
  'competition': 'Premier League',
  'age': 22},
 {'player': 'Yunis Abdelhamid',
  'nation': 'MAR',
  'position': 'DF',
  'squad': 'Reims',
  'competition': 'Ligue 1',
  'age': 35},
 {'player': 'Himad Abdelli',
  'nation': 'FRA',
  'position': 'MFFW',
  'squad': 'Angers',
  'competition': 'Ligue 1',
  'age': 23},
 {'player': 'Salis Abdul Samed',
  'nation': 'GHA',
  'position': 'MF',
  'squad': 'Lens',
  'competition': 'Ligue 1',
  'age': 22},
 {'player': 'Laurent Abergel',
  'nation': 'FRA',
  'position': 'MF',
  'squad': 'Lorient',
  'competition': 'Ligue 1',
  'age': 30},
 {'player': 'Oliver Abildgaard',
  'nation': 'DEN',
  'position': 'MF',
  'squad': 'Hellas Verona',
  'competition': 'Serie A',
  'age': 26},
 {'player': 'Matthis Abline',
  'nation': 'FRA',
  'position': 'FWMF',
  'squad': 'Auxerre',
  'competition': 'Ligue 1',
  'age': 19},
 {'player': 'Matthis Abline',
  'nation':

### 8. Create a new variable named players_by_nation that groups players by their respective nations.

The variable should be a dictionary where the keys are the nation names, and the values are lists of players belonging to that nation.

Each player should be represented as a dictionary containing the keys:

- player
- position
- squad
- competition
- age.

Example of expected output:

<code>
{'USA': [{'player': 'Brenden Aaronson',
          'position': 'MFFW',
          'squad': 'Leeds United',
          'competition': 'Premier League',
          'age': 31},
         {'player': 'Tyler Adams',
          'position': 'MF',
          'squad': 'Leeds United',
          'competition': 'Premier League',
          'age': 31},
          ...
        ],
        ...
}
</code>

In [27]:
players_by_nation = {}
for player in players:
    nation = player[1]
   
    player_data = {
        'player' : player[0],
        'position' : player[2],
        'squad' : player[3],
        'competition' : player[4],
        'age' : player[5]
    }

    if nation not in players_by_nation:
        players_by_nation[nation] = [player_data]
    else:
        players_by_nation[nation].append(player_data)

players_by_nation



{'USA': [{'player': 'Brenden Aaronson',
   'position': 'MFFW',
   'squad': 'Leeds United',
   'competition': 'Premier League',
   'age': 22},
  {'player': 'Tyler Adams',
   'position': 'MF',
   'squad': 'Leeds United',
   'competition': 'Premier League',
   'age': 23},
  {'player': 'John Brooks',
   'position': 'DF',
   'squad': 'Hoffenheim',
   'competition': 'Bundesliga',
   'age': 30},
  {'player': 'Timothy Chandler',
   'position': 'DF',
   'squad': 'Eint Frankfurt',
   'competition': 'Bundesliga',
   'age': 32},
  {'player': 'Sergiño Dest',
   'position': 'DF',
   'squad': 'Milan',
   'competition': 'Serie A',
   'age': 22},
  {'player': 'Luca de la Torre',
   'position': 'MF',
   'squad': 'Celta Vigo',
   'competition': 'La Liga',
   'age': 24},
  {'player': 'Weston McKennie',
   'position': 'MF',
   'squad': 'Leeds United',
   'competition': 'Premier League',
   'age': 24},
  {'player': 'Weston McKennie',
   'position': 'MFDF',
   'squad': 'Juventus',
   'competition': 'Serie A'

### 9. Create a new variable named players_by_squad that groups players by their respective squads.

The variable should be a dictionary where the keys are the squad names, and the values are lists of players belonging to that squad.

Each player should be represented as a dictionary containing the keys:

- player
- position
- squad
- competition
- age.

Example of expected output:
<code>
{'Leeds United': [{'player': 'Brenden Aaronson',
                    'position': 'MFFW',
                    'squad': 'Leeds United',
                    'competition': 'Premier League',
                    'age': 22},
                  {'player': 'Tyler Adams',
                   'position': 'MF',
                   'squad': 'Leeds United',
                   'competition': 'Premier League',
                   'age': 23},
                    ...
                ],
                ...
}
</code>

In [28]:
players_by_squad = {}

for player in players:
    squad = player[3]
    player_data = {
        'player': player[0],
        'position': player[2],
        'squad': player[3],
        'competition': player[4],
        'age': player[5]
    }

    if squad in players_by_squad:
        players_by_squad[squad].append(player_data)
    else:
        players_by_squad[squad] = [player_data]

players_by_squad

{'Leeds United': [{'player': 'Brenden Aaronson',
   'position': 'MFFW',
   'squad': 'Leeds United',
   'competition': 'Premier League',
   'age': 22},
  {'player': 'Tyler Adams',
   'position': 'MF',
   'squad': 'Leeds United',
   'competition': 'Premier League',
   'age': 23},
  {'player': 'Luke Ayling',
   'position': 'DF',
   'squad': 'Leeds United',
   'competition': 'Premier League',
   'age': 31},
  {'player': 'Patrick Bamford',
   'position': 'FW',
   'squad': 'Leeds United',
   'competition': 'Premier League',
   'age': 29},
  {'player': 'Liam Cooper',
   'position': 'DF',
   'squad': 'Leeds United',
   'competition': 'Premier League',
   'age': 31},
  {'player': 'Cody Drameh',
   'position': 'DF',
   'squad': 'Leeds United',
   'competition': 'Premier League',
   'age': 21},
  {'player': 'Mateo Fernández',
   'position': 'FWMF',
   'squad': 'Leeds United',
   'competition': 'Premier League',
   'age': 19},
  {'player': 'Junior Firpo',
   'position': 'DF',
   'squad': 'Leeds Un

### 10. Calculate the average age of players for each competition.

Create a new variable named average_age_by_competition that contains the average age for each competition. The competitions should be the keys, and the values should be the average ages rounded to one decimal place.

Example of expected output:

<code>
{
    'Premier League': 27.5,
    'La Liga': 26.8,
    'Bundesliga': 25.9,
    ...
}
</code>

* Note that, the value for age is rounded to the nearest 1 decimal point

In [30]:
from numpy import mean


average_age_by_competition = {}

for player in players:
    competition = player[4]
    age = player[5]

    if competition in average_age_by_competition:
        average_age_by_competition[competition].append(age)
    else:
        average_age_by_competition[competition] = [age]

for competition, ages in average_age_by_competition.items():
    average_age_by_competition[competition] = round(mean(ages), 1)

average_age_by_competition

{'Premier League': 26.0,
 'Ligue 1': 25.3,
 'Serie A': 26.0,
 'La Liga': 26.9,
 'Bundesliga': 25.8}

### 11. Transform the players data to dict-of-dict.

Transform the dataset to create a new variable named average_stats_by_position. This variable should contain the average values of goals, assists, and shots on target (SoT) for each position across all players. The positions should be the keys, and the values should be dictionaries with the keys goals, assists, and sot, representing the average values for each statistic.

Example of expected output:

<code>
{
    'FW': {'goals': 12.5, 'assists': 5.3, 'sot': 8.1},
    'MF': {'goals': 6.2, 'assists': 8.7, 'sot': 3.9},
    'DF': {'goals': 2.8, 'assists': 3.1, 'sot': 1.6},
    ...
}
</code>

* Note that, the values for goals, assists, sot are rounded to the nearest 1 decimal point

In [ ]:
position_stats = {}

for player in players:
    name, nation, pos, squad, comp, *rest, goals, shots, sot, assists = player
    goals = float(goals) 
    assists = float(assists)
    shots = float(shots)
    sot = float(sot)

    # Check if the position is already in the 'position_stats' dictionary
    if pos in position_stats:
        position_stats[pos]['goals'] += goals
        position_stats[pos]['assists'] += assists
        position_stats[pos]['shots'] += shots
        position_stats[pos]['sot'] += sot
        position_stats[pos]['count'] += 1
    else:
        # If the position is not in the 'position_stats' dictionary, add it
        position_stats[pos] = {
            'goals': goals,
            'assists': assists,
            'shots': shots,
            'sot': sot,
            'count': 1
        }


# Calculate the average stats for each position
average_stats_by_position = {}
for pos, stats in position_stats.items():
    average_stats_by_position[pos] = {
        'goals': round(stats['goals'] / stats['count'], 1),
        'assists': round(stats['assists'] / stats['count'], 1),
        'sot': round(stats['sot'] / stats['count'], 1)
    }


average_stats_by_position

{'MFFW': {'goals': 1.5, 'assists': 0.1, 'sot': 0.6},
 'DF': {'goals': 0.4, 'assists': 0.0, 'sot': 0.1},
 'MF': {'goals': 0.8, 'assists': 0.1, 'sot': 0.4},
 'FWMF': {'goals': 2.0, 'assists': 0.1, 'sot': 0.8},
 'FW': {'goals': 2.6, 'assists': 0.1, 'sot': 0.8},
 'DFFW': {'goals': 0.4, 'assists': 0.1, 'sot': 0.3},
 'MFDF': {'goals': 0.5, 'assists': 0.0, 'sot': 0.3},
 'GK': {'goals': 0.0, 'assists': 0.0, 'sot': 0.0},
 'DFMF': {'goals': 0.6, 'assists': 0.1, 'sot': 0.3},
 'FWDF': {'goals': 1.2, 'assists': 0.1, 'sot': 0.7}}

## Transformations/ Aggregations

### 12. Calculate total goals per Competition

Create a dictionary containing each competition as key, and the sum of all the goals scored as a value. Store the result in the variable goals_per_comp. It should look something like:

<code>
    {
        'Premier League': ...,
        'Serie A': ...,
        ...
        'Bundesliga': ...
    }
</code>

In [ ]:
goals_per_comp = {}
for player in players:
    name, nation, pos, squad, comp, *rest, goals, shots, sot, assists = player
    goals = int(goals)
    goals_per_comp.setdefault(comp, 0)
    goals_per_comp[comp] += goals
goals_per_comp

{'Premier League': 564,
 'Ligue 1': 618,
 'Serie A': 538,
 'La Liga': 480,
 'Bundesliga': 563}

### 13. Calculate the total number of minutes played by each squad across all competitions.

Create a dictionary named total_minutes_by_squad where the keys are the squad names, and the values are the total minutes played by that squad.

Example of expected output:

<code>
{
    'Manchester United': 4578,
    'Real Madrid': 5123,
    'Bayern Munich': 3984,
    ...
}

</code>

In [ ]:
total_minuets_by_squad = {}
for player in players:
    name, nation, pos, squad, *rest = player
    minutes = player[-5]
    total_minuets_by_squad.setdefault(squad, 0)
    total_minuets_by_squad[squad] += minutes
total_minuets_by_squad

{'Leeds United': 20743,
 'Reims': 21502,
 'Angers': 21611,
 'Lens': 21755,
 'Lorient': 21751,
 'Hellas Verona': 20671,
 'Auxerre': 21603,
 'Rennes': 21754,
 'Betis': 20524,
 'Toulouse': 21731,
 'Roma': 21780,
 'Inter': 20739,
 'Nantes': 21675,
 'Sevilla': 20604,
 'Southampton': 21780,
 'Köln': 18690,
 'Fulham': 22687,
 'Dortmund': 19800,
 'Leverkusen': 19764,
 'Milan': 21735,
 'Bologna': 20731,
 'Cremonese': 20743,
 'Troyes': 21704,
 'Werder Bremen': 19723,
 'Valladolid': 19791,
 'Spezia': 21720,
 'West Ham': 21780,
 'Monaco': 22554,
 'Crystal Palace': 21689,
 'Stuttgart': 19719,
 'Strasbourg': 21595,
 'Celta Vigo': 19708,
 'Torino': 21780,
 'Wolves': 21611,
 'Brentford': 21780,
 'Mainz 05': 19750,
 'Manchester City': 20725,
 'Almería': 20782,
 'Empoli': 21653,
 'Hoffenheim': 19724,
 'Real Madrid': 19799,
 'Cádiz': 20690,
 'Eint Frankfurt': 18787,
 'Barcelona': 19725,
 'Lazio': 21695,
 'Villarreal': 19765,
 'Leicester City': 21780,
 'Liverpool': 19766,
 'Getafe': 19782,
 'Elche': 19520

### 14. Determine the average number of starts for players in each competition.

Create a dictionary named average_starts_per_comp where the keys are the competition names, and the values are the average number of starts rounded to the nearest whole number.

Example of expected output:

<code>
{
    'Premier League': 23,
    'La Liga': 19,
    'Bundesliga': 21,
    ...
}
</code>

In [ ]:
starts_by_comp = {}
for player in players:
    name, nation, pos, squad, comp, age, born, mp, starts, *rest = player
    starts_by_comp.setdefault(comp, []).append(int(starts))

average_starts_per_comp = {}
for comp, starts_list in starts_by_comp.items():
    average_starts = round(sum(starts_list) / len(starts_list))
    average_starts_per_comp[comp] = average_starts

average_starts_per_comp

{'Premier League': 9,
 'Ligue 1': 9,
 'Serie A': 9,
 'La Liga': 8,
 'Bundesliga': 8}

### 15. Calculate the max scorers per competition.

Find the top scorers of each competition (maximum number of goals scored). Store your results in the variable top_scorers_per_comp. Attention! There might be more than one top scorer in the league, so your result should be a list of dictionaries containing each player and their goals as a tuple. 

Example:

<code>
{
    "Ligue 1": [
        ("Lionel Messi", 14),
        ("Kylian Mbappe", 14),
    ],
    'La Liga': [
        ('Robert Lewandowski', 18)
    ]
}
</code>

In [ ]:
top_scorers_per_comp = {}
for player in players:
    name,  nation, pos, squad, comp, *rest = player
    goals = player[-4]
    if comp not in top_scorers_per_comp:
        top_scorers_per_comp[comp] = [(name, goals)]
        continue
    current_scorers = top_scorers_per_comp[comp]
    current_top_goals = current_scorers[0][1]
    if goals > current_top_goals:
        top_scorers_per_comp[comp] = [(name, goals)]
    elif goals == current_top_goals:
        top_scorers_per_comp[comp].append((name, goals))
top_scorers_per_comp

{'Premier League': [('Erling Haaland', 25)],
 'Ligue 1': [('Folarin Balogun', 14), ('Wissam Ben Yedder', 14)],
 'Serie A': [('Victor Osimhen', 16)],
 'La Liga': [('Robert Lewandowski', 14)],
 'Bundesliga': [('Niclas Füllkrug', 13)]}

### 16. Convert the dataset into a new format - List of Dictionaries.

Create a list named goals_and_assists_by_player which is a list of dictionaries. Each dictionary should contain the keys player, competition, goals, and assists.

Example of expected output:

<code>
[
    {'player': 'Harry Kane', 'competition': 'Premier League', 'goals': 25, 'assists': 12},
    {'player': 'Lionel Messi', 'competition': 'La Liga', 'goals': 30, 'assists': 18},
    {'player': 'Cristiano Ronaldo', 'competition': 'Serie A', 'goals': 27, 'assists': 10},
    ...
]
</code>

In [ ]:
goals_and_assists_by_player = []

for player in players:
    name, nation, pos, squad, comp, age, born, mp, starts, mins, goals, shots, sot, assists = player
    goals_and_assists_by_player.append({
        'player': name,
        'competition': comp,
        'goals': goals,
        'assists': assists
    })

goals_and_assists_by_player

[{'player': 'Brenden Aaronson',
  'competition': 'Premier League',
  'goals': 1,
  'assists': 0.11},
 {'player': 'Yunis Abdelhamid',
  'competition': 'Ligue 1',
  'goals': 0,
  'assists': 0.05},
 {'player': 'Himad Abdelli',
  'competition': 'Ligue 1',
  'goals': 0,
  'assists': 0.0},
 {'player': 'Salis Abdul Samed',
  'competition': 'Ligue 1',
  'goals': 1,
  'assists': 0.0},
 {'player': 'Laurent Abergel',
  'competition': 'Ligue 1',
  'goals': 0,
  'assists': 0.08},
 {'player': 'Oliver Abildgaard',
  'competition': 'Serie A',
  'goals': 0,
  'assists': 0.0},
 {'player': 'Matthis Abline',
  'competition': 'Ligue 1',
  'goals': 1,
  'assists': 0.0},
 {'player': 'Matthis Abline',
  'competition': 'Ligue 1',
  'goals': 1,
  'assists': 0.0},
 {'player': 'Abner', 'competition': 'La Liga', 'goals': 0, 'assists': 0.0},
 {'player': 'Zakaria Aboukhlal',
  'competition': 'Ligue 1',
  'goals': 5,
  'assists': 0.24},
 {'player': 'Tammy Abraham',
  'competition': 'Serie A',
  'goals': 6,
  'assists

### 17. Calculate the total number of goals and assists for each player in the dataset

Expected Output is a dictionary with the player name as the key and the value is a dictionary with keys goals and assists and their respective values.

Example of expected output:

<code>
{
    'Harry Kane': {'goals': 25, 'assists': 12},
    'Lionel Messi': {'goals': 30, 'assists': 18},
    'Cristiano Ronaldo': {'goals': 27, 'assists': 10},
    ...
}
</code>

In [ ]:
goals_and_assists_per_player = {}

for player in players:
    name, nation, pos, squad, comp, age, born, mp, starts, mins, goals, shots, sot, assists = player
    goals = float(goals)
    assists = float(assists)
    if name not in goals_and_assists_per_player:
        goals_and_assists_per_player[name] = {'goals': goals, 'assists': assists}
    else:
        goals_and_assists_per_player[name]['goals'] += goals
        goals_and_assists_per_player[name]['assists'] += assists


goals_and_assists_per_player

{'Brenden Aaronson': {'goals': 1.0, 'assists': 0.11},
 'Yunis Abdelhamid': {'goals': 0.0, 'assists': 0.05},
 'Himad Abdelli': {'goals': 0.0, 'assists': 0.0},
 'Salis Abdul Samed': {'goals': 1.0, 'assists': 0.0},
 'Laurent Abergel': {'goals': 0.0, 'assists': 0.08},
 'Oliver Abildgaard': {'goals': 0.0, 'assists': 0.0},
 'Matthis Abline': {'goals': 2.0, 'assists': 0.0},
 'Abner': {'goals': 0.0, 'assists': 0.0},
 'Zakaria Aboukhlal': {'goals': 5.0, 'assists': 0.24},
 'Tammy Abraham': {'goals': 6.0, 'assists': 0.17},
 'Francesco Acerbi': {'goals': 0.0, 'assists': 0.08},
 'Mohamed Achi': {'goals': 0.0, 'assists': 0.0},
 'Marcos Acuña': {'goals': 2.0, 'assists': 0.09},
 'Che Adams': {'goals': 4.0, 'assists': 0.11},
 'Tyler Adams': {'goals': 0.0, 'assists': 0.0},
 'Sargis Adamyan': {'goals': 1.0, 'assists': 0.19},
 'Tosin Adarabioyo': {'goals': 1.0, 'assists': 0.0},
 'Martin Adeline': {'goals': 0.0, 'assists': 0.0},
 'Karim Adeyemi': {'goals': 2.0, 'assists': 0.12},
 'Amine Adli': {'goals': 1.

### 18. Groups players into different age groups

Create a new variable named players_by_age_group that groups players into different age groups. The variable should be a dictionary where the keys are the age group names (e.g., 'Under 20', '20-25', '26-30', 'Over 30'), and the values are lists of players belonging to that age group. Each player should be represented as a dictionary containing the keys player, nation, position, squad, competition, and age.

The age group is considered as:

Under 20 i.e <20
20-25 i.e 20 <= age <= 25
26-30 i.e 26 <= age <=30
Over 30 i.e >30
Example of expected output:

<code>
{
    'Under 20': [
        {'player': 'Player A', 'nation': 'Country A', 'position': 'Position A', 'squad': 'Squad A', 'competition': 'Competition A', 'age': 19},
        {'player': 'Player B', 'nation': 'Country B', 'position': 'Position B', 'squad': 'Squad B', 'competition': 'Competition B', 'age': 18},
        ...
    ],
    '20-25': [
        {'player': 'Player C', 'nation': 'Country C', 'position': 'Position C', 'squad': 'Squad C', 'competition': 'Competition C', 'age': 22},
        {'player': 'Player D', 'nation': 'Country D', 'position': 'Position D', 'squad': 'Squad D', 'competition': 'Competition D', 'age': 25},
        ...
    ],
    ...
}
</code>

In [ ]:
players_by_age_group = {
    'Under 20': [],
    '20-25': [],
    '26-30': [],
    'Over 30': []
}

for player in players:
    name, nation, pos, squad, comp, age, *rest = player
    age = int(age)
    if age < 20:
        age_group = 'Under 20'
    elif 20 <= age <= 25:
        age_group = '20-25'
    elif 26 <= age <= 30:
        age_group = '26-30'
    else:
        age_group = 'Over 30'

    players_by_age_group[age_group].append({
        'player': name,
        'nation': nation,
        'position': pos,
        'squad': squad,
        'competition': comp,
        'age': age
    })


players_by_age_group

{'Under 20': [{'player': 'Matthis Abline',
   'nation': 'FRA',
   'position': 'FWMF',
   'squad': 'Auxerre',
   'competition': 'Ligue 1',
   'age': 19},
  {'player': 'Matthis Abline',
   'nation': 'FRA',
   'position': 'FWMF',
   'squad': 'Rennes',
   'competition': 'Ligue 1',
   'age': 19},
  {'player': 'Martin Adeline',
   'nation': 'FRA',
   'position': 'MFFW',
   'squad': 'Reims',
   'competition': 'Ligue 1',
   'age': 19},
  {'player': 'Amar Ahmed',
   'nation': 'SWE',
   'position': 'FW',
   'squad': 'Troyes',
   'competition': 'Ligue 1',
   'age': 18},
  {'player': 'Mohamed Ali Cho',
   'nation': 'FRA',
   'position': 'FW',
   'squad': 'Real Sociedad',
   'competition': 'La Liga',
   'age': 19},
  {'player': 'Carlos Álvarez',
   'nation': 'ESP',
   'position': 'FW',
   'squad': 'Sevilla',
   'competition': 'La Liga',
   'age': 19},
  {'player': 'Taïryk Arconte',
   'nation': 'FRA',
   'position': 'FWMF',
   'squad': 'Brest',
   'competition': 'Ligue 1',
   'age': 19},
  {'player